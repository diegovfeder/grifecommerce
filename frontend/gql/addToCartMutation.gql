	mutation AddToCartMutation($id: ID!) {
		addToCart(productId: $id) {
			id
		}
	}
