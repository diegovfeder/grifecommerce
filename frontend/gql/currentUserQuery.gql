query CurrentUser {
		authenticatedItem {
			... on User {
				id
				email
				name
				cart {
					id
					quantity
					product {
						id
						price
						name
						description
						photo {
							image {
								publicUrlTransformed
							}
						}
					}
				}
			}
		}
	}
