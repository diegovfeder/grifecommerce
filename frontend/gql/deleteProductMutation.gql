mutation DELETE_PRODUCT_MUTATION($id: ID!) {
	deleteProduct(id: $id) {
		id
		name
	}
}
