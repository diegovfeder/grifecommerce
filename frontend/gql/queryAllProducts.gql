query AllProducts {
	products {
		id
		name
		price
		description
		photo {
			id
			image {
				publicUrlTransformed
			}
		}
	}
}
