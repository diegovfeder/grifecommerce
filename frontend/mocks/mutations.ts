import DELETE_PRODUCT_MUTATION from '../gql/deleteProductMutation.gql';
import ADD_TO_CART_MUTATION from '../gql/addToCartMutation.gql';

const mockedMutations = {
	addToCart: {
		request: {
			query: ADD_TO_CART_MUTATION,
			variables: {
				id: '1',
			},
		},
		result: {
			data: {
				addToCart: {
					id: '1',
				},
			},
		},
	},
	deleteProduct: {
		request: {
			query: DELETE_PRODUCT_MUTATION,
			variables: {
				id: '1',
			},
		},
		result: {
			data: {
				deleteProduct: {
					id: '1',
					name: 'Sample Pack #01',
				},
			},
		},
	},
	signUp: {
		// TODO:
	},
};

export default mockedMutations;
