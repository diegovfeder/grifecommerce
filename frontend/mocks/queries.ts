import { CURRENT_USER_QUERY } from '../hooks/useUserQuery';

const mockQueries = {
	currentUser: {
		request: {
			query: CURRENT_USER_QUERY,
			variables: {},
		},
		result: {
			data: {
				authenticatedItem: {
					id: '1',
					email: 'diego@test.com',
					name: 'Diego Test',
					cart: {
						id: '1',
						quantity: '1',
						product: {
							id: '1',
							price: '200',
							name: 'Sample Pack #01',
							description: 'This is a sample pack',
							photo: {
								image: {
									publicUrlTransformed: 'https://via.placeholder.com/300x300',
								},
							},
						},
					},
				},
			},
		},
	},
};

export default mockQueries;
