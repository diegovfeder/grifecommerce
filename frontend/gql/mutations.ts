import gql from 'graphql-tag';

export const SEND_USER_PASSWORD_RESET_LINK_MUTATION = gql`
	mutation SendUserPasswordResetLink($email: String!) {
		sendUserPasswordResetLink(email: $email)
	}
`;

export const SIGN_UP_MUTATION = gql`
	mutation SIGNUP_MUTATION(
		$email: String!
		$name: String!
		$password: String!
	) {
		createUser(data: { email: $email, name: $name, password: $password }) {
			id
			email
			name
		}
	}
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CreateOrder($token: String!) {
		checkout(token: $token) {
			id
			charge
			total
			items {
				id
				name
			}
		}
	}
`;
