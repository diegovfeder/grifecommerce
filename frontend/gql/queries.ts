import gql from 'graphql-tag';

export const USER_EMAIL_QUERY = gql`
	query UserEmail($email: String!) {
		user(where: { email: $email }) {
			id
			email
		}
	}
`;

export const PRODUCT_QUERY = gql`
	query Product($id: ID!) {
		product(where: { id: $id }) {
			id
			name
			price
			description
			photo {
				id
				altText
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const ALL_PRODUCTS_QUERY = gql`
	query AllProducts {
		products {
			id
			name
			price
			description
			photo {
				id
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const PRODUCTS_QUERY = gql`
	query Products($take: Int, $skip: Int = 0) {
		products(take: $take, skip: $skip) {
			id
			name
			price
			description
			photo {
				id
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const PRODUCTS_COUNT_QUERY = gql`
	query ProductsCount {
		productsCount
	}
`;

export const CURRENT_USER_QUERY = gql`
	query CurrentUser {
		authenticatedItem {
			... on User {
				id
				email
				name
				cart {
					id
					quantity
					product {
						id
						price
						name
						description
						photo {
							image {
								publicUrlTransformed
							}
						}
					}
				}
			}
		}
	}
`;
