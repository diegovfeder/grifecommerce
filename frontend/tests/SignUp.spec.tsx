import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { resolveMockState } from './utils';
import { SIGN_UP_MUTATION } from '../gql/mutations';
import { CURRENT_USER_QUERY } from '../gql/queries';
import Signup from '../components/SignUp';

const mocks: MockedResponse[] = [
	{
		request: {
			query: SIGN_UP_MUTATION,
			variables: {
				name: '',
				email: '',
				password: '',
			},
		},
		result: {
			data: {
				createUser: null,
			},
		},
		error: new Error('Variables were not provided.'),
	},
	{
		request: {
			query: SIGN_UP_MUTATION,
			variables: {
				name: 'Diego Feder',
				email: 'diegovfeder@gmail.com',
				password: '12345678',
			},
		},
		result: {
			data: {
				createUser: {
					id: 'b567b953-2dab-4553-abc8-fdc2a55e3bb2',
					email: 'diegovfeder@gmail.com',
					name: 'Diego Feder',
				},
			},
		},
	},
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: {
				authenticatedItem: {
					id: 'b567b953-2dab-4553-abc8-fdc2a55e3bb2',
					email: 'diegovfeder@gmail.com',
					name: 'Diego Feder',
					cart: [],
				},
			},
		},
	},
];

describe('<SignUp/>', () => {
	it('renders component and matches snapshot', () => {
		const { container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Signup />
			</MockedProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	it('renders without errors', async () => {
		render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Signup />
			</MockedProvider>,
		);
		expect(screen.getByText('Sign Up For an Account')).toBeVisible();
		expect(screen.getByText('Name')).toBeVisible();
		expect(screen.getByText('Email')).toBeVisible();
		expect(screen.getByText('Password')).toBeVisible();
		expect(screen.getByText('Sign Up!')).toBeVisible();
		expect(screen.queryByTestId('error-message-component')).toBe(null);
		expect(screen.queryByTestId('graphql-error')).toBe(null);
		expect(screen.queryByText('Error:')).toBe(null);
		expect(screen.queryByText('Variables were not provided.')).toBe(null);
	});

	describe('when all inputs are empty', () => {
		describe('when the user clicks on the sign up button', () => {
			it('calls the mutation, gets null as return and renders error', async () => {
				render(
					<MockedProvider mocks={mocks} addTypename={false}>
						<Signup />
					</MockedProvider>,
				);
				expect(screen.getByText('Sign Up For an Account')).toBeVisible();
				expect(screen.getByText('Name')).toBeVisible();
				expect(screen.getByText('Email')).toBeVisible();
				expect(screen.getByText('Password')).toBeVisible();
				expect(screen.getByText('Sign Up!')).toBeVisible();
				expect(screen.queryByTestId('error-message-component')).toBe(null);
				expect(screen.queryByTestId('graphql-error')).toBe(null);
				await userEvent.click(screen.getByText('Sign Up!'));
				await resolveMockState();
				expect(screen.queryByTestId('error-message-component')).toBe(null);
				expect(screen.queryByTestId('graphql-error')).toBe(null);
				expect(screen.getByText('Please fill in all fields.')).toBeVisible();
			});
		});
	});

	describe('when all inputs are filled', () => {
		describe('when the user clicks on the sign up button', () => {
			it('calls the mutation properly', async () => {
				render(
					<MockedProvider mocks={mocks} addTypename={false}>
						<Signup />
					</MockedProvider>,
				);
				expect(screen.getByText('Sign Up For an Account')).toBeVisible();
				expect(screen.getByLabelText(/name/i)).toBeInTheDocument();
				expect(screen.getByText('Name')).toBeVisible();
				expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
				expect(screen.getByText('Email')).toBeVisible();
				expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
				expect(screen.getByText('Password')).toBeVisible();
				expect(screen.getByRole('button')).toBeVisible();
				expect(screen.getByText('Sign Up!')).toBeVisible();
				await userEvent.type(screen.getByPlaceholderText(/name/i), 'Diego Feder');
				await userEvent.type(screen.getByPlaceholderText(/email/i), 'diegovfeder@gmail.com');
				await userEvent.type(screen.getByPlaceholderText(/password/i), '12345678');
				await userEvent.click(screen.getByText('Sign Up!'));
				await resolveMockState();
				expect(screen.queryByText('Please fill in all fields.')).toBe(null);
				await screen.findByText(
					`Signed Up with diegovfeder@gmail.com - Please Go Ahead and Sign In!`,
				);
			});
		});
	});
});
