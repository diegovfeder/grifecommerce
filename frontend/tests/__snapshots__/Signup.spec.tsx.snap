// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> renders component and matches snapshot 1`] = `
<div>
  <form
    autocomplete="new-password"
    class="StyledForm-sc-44wenk-0 iUkmmK"
    method="POST"
  >
    <h2>
      Sign Up For an Account
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Name
        <input
          aria-label="name"
          autocomplete="new-password"
          name="name"
          placeholder="Your Name"
          type="text"
          value=""
        />
      </label>
      <label
        for="email"
      >
        Email
        <input
          aria-label="email"
          autocomplete="new-password"
          name="email"
          placeholder="Your Email Address"
          type="email"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          aria-label="password"
          autocomplete="new-password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign Up!
      </button>
    </fieldset>
  </form>
</div>
`;
